### PR TITLE
Support NDP Proxying

### PR DESCRIPTION
**Pull request by furkansahin**
We are running on a landscape that different hosting providers have different implementations on how they redirect traffic for IPv6 ranges.
    There are 2 main themes;
    1. For a given subnet, redirect the whole traffic according to prefix by skipping ndp. In this implementation, if we are assigned /64 subnet, all the packets in that range will be sent to the host.
    2. We are assigned /64 subnet, we split it to /79 chunks but when we need to send a packet to that /79 chunk, switch sends ns messages to resolve the MAC address. If it doesn't get any response, the package is dropped. That's why our host needs to respond to these messages via proxy ndp and we should teach it to respond a specific ip address even if it is under our range. This comes with a cost of not being able to use all the ip addresses in the range since we have to add them 1 by 1 but it is a hosting provider limitation and we cannot do much atm.

### CODE REVIEW

**Comment by fdr** [comment]
Let's change the commit message. I don't think proxy_ndp will be unique to Mars by any stretch, given what I dug about OVHCloud as well. I think the jargon would be closer to: "Support NDP proxying" / "For networks that..."

Now the "..." does hide some of my confusion of terminology, because there are several related terms to being able to receive a prefix's packets without NDP, such as: static routing, or prefix delegation, but both of these terms include meanings that we don't particularly care about.

So you may need to describe the situation in some detail rather than being able to use a few jargon words (that I know of, right now)

**Comment by enescakir** [comment]
Screenshot 2023-06-05 at 14 56 25

File: migrate/011_ipv6_ndp_support.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:vm_host) do
+      # This will be moved to a separate table
+      # once ipv4 is supported
```
Comment (fdr) [inline migrate/011_ipv6_ndp_support.rb:7]:
```suggestion
```

Not sure if it will or won't be practical, let's leave it a mystery.

File: prog/install_dnsmasq.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ def compile_and_install
   end
 
   def install_build_dependencies
-    sshable.cmd("sudo apt-get -y install make gcc")
+    sshable.cmd("sudo apt-get -y install make gcc git")
```
Comment (fdr) [inline prog/install_dnsmasq.rb:28]: This seems like an unrelated change, what's up here? Maybe a difference in base image preparation? Does it belong in a separate commit?

File: rhizome/bin/prep_host.rb [hunk 3]
```diff
@@ -36,3 +37,6 @@
 # For qemu-image convert and mcopy for cloud-init with the nocloud
 # driver.
 r "apt-get -y install qemu-utils mtools"
+
+# For some providers like Mars
+r "apt-get install -y dmidecode dosfstools"
```
Comment (fdr) [inline rhizome/bin/prep_host.rb:42]: I suppose his is also a case of "packages that are not installed by default...sometimes". And it's probably not a mars-ism so much as how ubuntu is prepared on the server, since IIRC we're the ones who prepared it from the upstream iso in this case. Perhaps should have a "add missing dependencies for minimal Ubuntu installs" commit.

File: rhizome/lib/vm_setup.rb [hunk 4]
```diff
@@ -92,7 +92,8 @@ def subdivide_network(net)
     [halved, halved.next_sib]
   end
 
-  def routes(gua, private_subnets)
+  def routes(gua, private_subnets, ndp_needed)
+    main_device = r "ip route | awk '/default/ {print $5}'"
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:96]: This is admirably terse, but I wonder if it is as sound as `ip -j route` would be. Is the latter _also_ short, or short-ish? It would be a surplus in soundness if we could do it that way.

**Review by fdr — CHANGES_REQUESTED** [review]
@enescakir I presume you had a had in the checkbox.erb file? It's kind of a new feature tucked in this patch, to add boolean flags to the interface in general.

File: prog/install_dnsmasq.rb (hunk 2 above)
Comment (furkansahin) [reply prog/install_dnsmasq.rb:28]: it was needed since we did install minimal ubuntu but now I set them up with the full image, so this shouldn't be necessary.

File: rhizome/bin/prep_host.rb (hunk 3 above)
Comment (furkansahin) [reply rhizome/bin/prep_host.rb:42]: should I move it to a different PR?

File: rhizome/lib/vm_setup.rb (hunk 4 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: here you go

**Review by fdr — APPROVED** [review]

**Comment by fdr** [comment]
There's a few changes in here unrelated to NDP proxying (e.g. shuffling the product line)  that I'd prefer to be separate somehow, but I suppose that may be too pedantic.
